src/wallet/coin_selection.rs

```diff
@@ -534,6 +534,49 @@ impl BranchAndBoundCoinSelection {
     }
 }
 
+struct WasteMetric;
```

Comment:
This looks like a good direction for improving coin selection, I haven't looked deeply into the code/algorithm but I do have a couple rust style suggestions.

1. Unless you plan on adding additional fields to the `WasteMetric` struct I think you can do without it.  
2. And if the idea for `calculate_waste` is as a helper function to implementations of the `coin_select` trait then it may be better to make `calculate_waste` a "default implementation" on the `coin_selection` trait.

Reply:
The waste metric requires a long term fee rate (expected average fee rate of the wallet) so idk if it can be directly be added to `CoinSelectionResult`.

Reply:
Just to clarify, the "Long Term Feerate Estimate" (LTFRE) is not the average feerate on the wallet, it's more of a minimum feerate that you expect to be able to spend at in the future. The thinking is that if you expect to be able to spend an input at a lower feerate in the future, you should delay spending it to then, hence, anything above that feerate is considered "wasteful". Since you always need to use at least one input, two solutions that each use one input of the same type will see their score increase by the same amount, so it doesn't affect their prioritization. However, if you e.g. have different output types or compare solutions with multiple inputs, you'll prefer the smaller input weight per this heuristic.